### DISCUSSION TITLE
Suffix to denote byte literals

### DISCUSSION THREAD

**Discussion by IanKemp** [dotnet/csharplang#6496] [Language Ideas]
From https://stackoverflow.com/questions/5378036/literal-suffix-for-byte-in-net and https://roslyn.codeplex.com/discussions/542111 - it seems this has been on the wishlist of quite a few devs for a long time, and there was talk of implementing it at the same time that the binary prefix was added, but that apparently never happened.

The end result is that every time I want to declare a byte literal, I have to manually cast it, which is (annoying|fugly|unnecessary|take-your-pick).

Is there any chance we'll see this in the C# language anytime soon, especially since (as per the linked CodePlex discussion) VB.NET has had it for more than 3 years?

**Comment by jnm2** [comment]
The asymmetry is the main thing that gets me. I really want `sb` for `sbyte` and `s` and `us` for `short` and `ushort` too.

This is why: https://github.com/jnm2/ApiContractGenerator/blob/29b9afb0cece604752c786bb3fd0334d37d809d4/src/ApiContractGenerator.Tests/CSharpTextFormatterTests.cs#L9-L70

It comes up not infrequently.

**Comment by iSazonov** [comment]
In PowerShell repo we want to implement a suffix for byte type. Usually we follow C# style. So it would be nice to have here some conclusion. 
Currently we consider 'y' because 'b' can mean 'bit'.

**Comment by 4creators** [comment]
This is a general deficiency of C# which spans essentially all numeric types. IMHO we should create consistent way to name literals representing all implemented numeral types and keep them open for expansion.

.NET has adapted very clear and communicative way of naming integers except for byte and sbyte.  Every integer name is self explanatory i.e. `Int32`, `Int64`, `UInt64`. It immediately and in a self explanatory way can be translated into literal suffixes:

```C#
var b = 0xf7u8;             // byte
var b2 = b0011001100u8;     // byte
var s = 0x07i8;             // sbyte
var sh = 0x0657i16;         // short
var us = 0xf543u16;         // ushort
/// etc. etc. etc.
var i = 0xfde2i128;         // Int128
```

But what is even more important it can be easily expanded into new numerical literals covering binary and decimal floats:

```C#

var h = 0.0163f16;             // half - IEEE:754 binary16
var b = 0.0163f128;            // IEEE:754 binary128
var e = 0.1f8;                 // non standardized 8 bit float

```

The same goes for decimals, complex, quaternions, octonions. The naming scheme is easy to understand for every programmer coming from any programming language and even easier to follow as it is based on very simple naming rules. Furthermore, there are plans to implement new numeric types due to the requirements of ML.NET or expanding hardware support. See propsal to implement `Half` and perhaps `Binary float` in this issue:  [API Proposal: Add System.Numerics.Half 16 bit floating point number conforming to IEEE 754:2008 binary16](https://github.com/dotnet/corefx/issues/25702).

IMHO we should not solve `byte` problem separately but discuss thoroughly numeric literals support in C#. Similar discussion may be looming on the horizon for string literals encodings as well.

**Reply by light-traveller** [reply]
I prefer shortest possible suffixes. My proposal as stated in https://github.com/dotnet/csharplang/discussions/6774 is `Y` and `y` for `byte` literals. Long suffixes are in contrast with the purpose of enabling the programmer to write shorter code and makes the language and code, bloated, cluttered and less readable. Also, being able to pick up on the type of the variable by a single easy look is absolutely preferred over trying to visually separate more letters at the end of the value and then mapping it to the type it denotes.

**Reply by tannergooding** [reply]
Only so many letters exist and many of the existing suffixes, such as L are already problematic for other reasons.

We realistically we need literals to cover byte, sbyte, short, and ushort. We may also in the future need suffixes to cover int128/uint128, half, 128-biy binary float, or more.

i# or s#, u#, and f# are pretty clear and unambiguous for any type of signed/unsigned integer, or floating-point data. We already settled on u8 for utf8 strings and those are naturally typed as ROSpan<byte> so likewise using it to mean "unsigned byte" in general makes a decent amount of sense.

That would give us u8, u16, u32, and u64 for unsigned integers, with a clear extension for u128 if the language gets official support.

Then f32 and f64 for floating point. With clear extensions for f16/f128

Then for signed integers i (for integer) or s (for signed) ends up working for thr same pattern.

These all are also inline with what other languages have done or are doing to avoid the kind of problems C/C++ had with picking letters that are highly conflicting or potentially ambiguous.

Even C/C++ 23 are adding support for some of these patterns to improve code readability and type guarantees

**Reply by light-traveller** [reply]
Reading something like `0x0163i128` is a real pain especially when there are numerous literals in a region of code and one might want to pay attention to the values for debugging purposes or understand the modification of values. In such situations, debugging is visually grueling.

Moreover, other numeric types (`ushort`, etc.) are seldom used compared to `byte` which is used quite frequently in byte manipulation scenarios. In my opinion, your proposal misses to account for practicality and programmers' ease in long term.

**Reply by jnm2** [reply]
You can made it more readable by doing `0x0163_i128`.

**Reply by light-traveller** [reply]
I can but inherently it is not

**Reply by CyrusNajmabadi** [reply]
Looks very readable and composable to me.  It would be readily apparent what this means to almost anyone coming to c#

**Reply by light-traveller** [reply]
@CyrusNajmabadi, Seems you get paid for expressing disagreement in whatever way possible in different threads!

**Reply by light-traveller** [reply]
Disagree

**Comment by DoctorLogiq** [comment]
I find that a suffix to write byte literals would be great. For example, I'm working with OpenTK (C#'s OpenGL bindings) and the function `GL.Color3( )` takes most numeric types, including `int` and `byte`. When I pass a byte value such as 255 to this function, it casts it to an int by default, meaning I have to explicitly cast them to bytes. It would be more convenient (easier to write, time and space-saving) to suffix the literals to say "this is a byte, not an integer". Also it would be helpful if such a literal can be forcibly clamped to its min-max values by the IDE/compiler etc.

(On another note) I don't know, maybe there should also be a normalised float type (suffixed with 'n' / 'nf'?) which only allows values between 0.0F and 1.0F, since normalisation is used in a damn lot of places. This would reduce the need for writing code to check those values are within the 0-1 range all over the place.

Thoughts? (Take into account I'm no compiler/language expert, so I understand there'll no doubt be reasons not to do these things, or at least reasons not to do them as I've suggested).

**Comment by dzmitry-lahoda** [comment]
This feature is must for type safe code. Also Rust support that.

**Comment by stan-sz** [comment]
> In PowerShell repo we want to implement a suffix for byte type. Usually we follow C# style. So it would be nice to have here some conclusion. Currently we consider 'y' because 'b' can mean 'bit'.

How about `b` for bit and `B` for byte?

**Reply by light-traveller** [reply]
`B` and `b` conflict with hexadecimal notation. e.g.: `0xFCB`. Is this a byte with value of `FC` or a hexadecimal with value of `FCB`?

**Comment by FaustVX** [comment]
Currently, you could already type the suffix both in lower-case and upper-case.
`123ul` equals `123UL` equals `123uL` equals `123Ul`

**Comment by bernd5** [comment]
For me this is not a big issue because we have constant-conversion, today you need to write for example `(byte)255`, instead of `255b`.
But I agree that a suffix notation would be nice - like we have them for other numeric types.

**Comment by light-traveller** [comment]
As I have stated in #6774, in my opinion 'y' & 'Y' are good choices for denoting byte literals.
